File: app/src/main/java/com/example/rssreader/presentation/feed/FeedPresenter.java
package com.example.rssreader.presentation.feed;

import android.net.Uri;
import android.support.annotation.NonNull;
import android.util.Log;

import com.example.rssreader.data.DataRetriever;

import kotlin.Unit;

public class FeedPresenter {
    private static final String TAG = "FeedPresenter";

    @NonNull private DataRetriever retriever;
    @NonNull private FeedView view;

    public FeedPresenter(@NonNull FeedView view) {
        this.view = view;
        retriever = new DataRetriever(Uri.parse("https://meduza.io/rss/all/"),
                (data) -> {
                    view.showData(data);
                    return Unit.INSTANCE;
                },
                (code, t) -> {
                    Log.i(TAG, "onError: code: " + code);
                    Log.i(TAG, "onError: error: " + t);
                    view.showError();

                    return Unit.INSTANCE;
                }) ;
    }

    public void requestData() {
        retriever.load();
    }

    public void stop() {
        retriever.cancel();
    }
}

File: app/src/main/java/com/example/rssreader/presentation/feed/FeedPresenter.kt
package com.example.rssreader.presentation.feed

import android.net.Uri
import com.example.rssreader.data.repository.RssItemRepository
import com.example.rssreader.di.RssServiceFactory
import com.example.rssreader.domain.entity.RssItem
import com.example.rssreader.logInfo
import com.example.rssreader.presentation.BasePresenter
import io.reactivex.android.schedulers.AndroidSchedulers

class FeedPresenter(private val view: FeedView) : BasePresenter() {
    private val itemRepository = RssItemRepository(RssServiceFactory.service)
    private lateinit var data: List<RssItem>

    override fun start() {
        load()
    }

    fun onRefresh() {
        stop()
        load()
    }

    fun onItemClick(position: Int) {
        view.openDetails(data[position].link)
    }

    private fun load() {
        itemRepository.loadAllFrom(Uri.parse("https://meduza.io/rss/all/"))
            .observeOn(AndroidSchedulers.mainThread())
            .doOnSubscribe { view.showProgress() }
            .doOnTerminate { view.hideProgress() }
            .subscribe(this::onSuccess, this::onError)
            .clearOnDestroy()
    }

    private fun onSuccess(data: List<RssItem>) {
        this.data = data
        view.showData(data)
    }

    private fun onError(t: Throwable) {
        logInfo("onError: error: $t", t)
        view.showError()
    }
}